h5netcdf/core.py

```diff
@@ -194,28 +207,53 @@ def _lookup_dimensions(self):
         attrs = self._h5ds.attrs
         # coordinate variable and dimension, eg. 1D ("time") or 2D string variable
         if (
-            "_Netcdf4Coordinates" in attrs
-            and attrs.get("CLASS", None) == b"DIMENSION_SCALE"
+                "_Netcdf4Coordinates" in attrs
+                and (
+                    attrs.get("CLASS", None) == b"DIMENSION_SCALE"
+                    or self._parent.backend == "pyfive"
+                )
         ):
-            order_dim = {
-                value._dimid: key for key, value in self._parent._all_dimensions.items()
-            }
+            # Note: For the pyfive backend, if "_Netcdf4Coordinates"
+            #       exist then we should use this method as it is much
+            #       faster than using the DIMENSION_LIST method
+            try:
+                # Try using cached order_dim
+                if "legacy" in self._cls_name:
+                    # Haven't yet worked how to do caching with the
+                    # legacy API (because as things stand we end up
+                    # with recursion in
+                    # HasAttributesMixin.__[gs]etattr__ when we try to
+                    # get/set the _cached_h5ds attribute).
+                    raise AttributeError
+
+                order_dim = self._parent._cached_order_dim
+            except AttributeError:
```

Comment:
What is going on. If you have a `try` with a `raise attributerror` and an `except attributerror`  around it?

Reply:
When the `raise AttributeError` happens, it jumps us to the `except AttributeError:` block, just as if a line of code had resulted in an AttributeError being raised.